### DISCUSSION TITLE
Parser combinators design patterns versus parsley Haskell 2.0.0.0

### DISCUSSION THREAD

**Discussion by funarog** [j-mie6/design-patterns-for-parser-combinators#10] [General]
I am having trouble implementing recommended parser combinator design patterns using parsley 2.0.0.0  An implementation of a simple Expr function using design pattern for white space and  yields a build error. Can't find a solution in any Parsley documentation.

The code compiles if function comments are removed

Design pattern:  

```haskell
whitespace :: Parser ()
whitespace = void (many (satisfy isSpace))

lexeme :: Parser a -> Parser a
lexeme p = p <* whitespace` 
```
function:
```haskell
expr :: Parsley.Parser Expr
expr = whitespace *> sum' <* eof
    where
        -- whitespace = skipMany (oneOf " \n")
        -- lexeme p = p <* whitespace
        operator op = lexeme $ string op

        digit =
            QQ (\c -> fromIntegral (ord c - ord '0'))
                <$> oneOf "0123456789"
        number = lexeme $ QQ (Num) <$> somel (QQ (\q r -> 10 * q + r)) (LIFTED 0) digit

        atom = lexeme (char '(') *> sum' <* lexeme (char ')') <|> number
        sum' =
            precHomo
                atom
                [ ops
                    InfixL
                    [ operator "*" $> QQ (Bin Mul)
                    , operator "/" $> QQ (Bin Div)
                    ]
                , ops
                    InfixL
                    [ operator "+" $> QQ (Bin Add)
                    , operator "-" $> QQ (Bin Sub)
                    ]
                ]
```
Build profile: -w ghc-9.2.7 -O1

> 
src/Parser.hs:61:26: error:
    • Couldn't match type ‘(->) Char’ with ‘Defunc’
        arising from a use of ‘satisfy’
    • In the first argument of ‘many’, namely ‘(satisfy isSpace)’
      In the first argument of ‘void’, namely ‘(many (satisfy isSpace))’
      In the expression: void (many (satisfy isSpace))
   |
61 | whitespace = void (many (satisfy isSpace))

Imports 
```haskell
{-# LANGUAGE CPP #-}
{-# LANGUAGE PackageImports #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE LambdaCase #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Unused LANGUAGE pragma" #-}
{-# HLINT ignore "Redundant bracket" #-}


#define QQ(x) (makeQ (x) [|| (x) ||])

'''

**Comment by j-mie6** [comment]
`QQ(isSpace)`!

**Comment by j-mie6** [comment]
As a side note, when I _finally_ get some time to deal with my huge backlog, I do intend to write a proper wiki/tutorial for parsley Haskell. And bring more of the "proper" API in from Scala parsley. We can do a _lot_ better.

**Comment by funarog** [comment]
Well that worked!   Thanks.


I am currently working with [haskell-parsing-benchmarks](https://gitlab.com/FinnBender/haskell-parsing-benchmarks) to try and determine if the claim that flatparse is the fastest haskell parser is true. I suspect that one is comparing apples with oranges. Nonetheless "[when [you] get some time ..." and "write a proper wiki/tutorial for parsley.: That would be great.   This gives me an idea to perhaps   benchmark Scala parsley.

**Reply by j-mie6** [reply]
Benchmarking Scala parsley again versus `cats-parse`, `fastparse` etc is something I've been wanting to get done for a while (the architecture is much improved since the last time it was benchmarked!)

It's a bit tricky though: ideally you make a distinction between "hot" and "cold" parsers: `parsley` Scala incurs overheads the first time the parser is used to compile it to the abstract machine (cold), and then every time after that it runs much faster (hot). The biggest improvement I can get at this point on that architecture is probably shaving a little more time off the cold booting for it.

**Reply by j-mie6** [reply]
If possible, it might be worth pulling the HEAD of `parsley-core` down for your benchmarks: I've made a bunch of improvements to `parsley-core` that I've not have a change to asses the "damage" for a new published version on. I use this in my internal benchmarks:

```
source-repository-package
    type: git
    location: https://github.com/j-mie6/ParsleyHaskell.git
    tag: 165f99d7e16d7e2f505425402cd0eabcbb496ea8
    subdir: parsley-core
```

**Reply by funarog** [reply]
Sure. I will give it a go.
